### PR TITLE
[dynamo] preserve some FX node metadata of GraphModules

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107067

Requested from @tugsbayasgalan: we want dynamo to preserve some FX node metadata when we trace `GraphModule`s (`nn_module_stack`, `source_fn`, `stack_trace`). This is helpful for the case when we export an aten-level `GraphModule`, add some (possibly non-torch or non-aten) ops, and we want to transform the graph back into an aten-level graph. Without preserving metadata, future passes that look at metadata (e.g. quantization passes) won't work.

This feature also has the additional benefit of being able to preserve origin line of code when `print_readable`'ing a `GraphModule`. This is helpful when debugging graphs that have passed through dynamo several times.

The added unit test demonstrates the added functionality of this PR.

~This PR is currently a proof-of-concept implementation that shows that preserving node metadata across dynamo is possible.~ This PR preserves node metadata across dynamo by doing the following:
- ~inject a counter variable into the `GraphModule` source code, which is incremented every time a node is run~
- Construct a line number -> node index map in `GraphModule` as the source code is being generated.
- pass a list of node metadata and the line number map to dynamo's bytecode analyzer
- ~dynamo traces the counter as a `ConstantVariable`, so when we create a new proxy, we can determine which original node index this proxy corresponds by looking at the value of the traced counter~
- When we create a new proxy, get the current instruction's line number, and get the node index using the line number map
- index into the original node metadata ~using the counter variable's tracked value.~

~Some things that should be addressed off the top of my head:~
- ~Is this feature even desirable? (Do we really want Dynamo to have special behavior for `GraphModules`? Should we expect users to re-export `GraphModules`?)~
- ~Is there a better approach than to use a counter? We considered using node names, line numbers, and assuming that proxies are created in the same order as the nodes, but each of these 3 have shortcomings. For node names, we only have access to new node names, not the old ones. Using line number is fragile. The third is problematic since not all created nodes go through `create_proxy` (e.g. inputs). We currently generate a line number to node index map when the `GraphModule`'s code is generated.~
- ~What's the best way to send data across the "CPython gap"? That is, it is not obvious how to cleanly pass data from dynamo's `eval_frame.py:_TorchDynamoContext.__call__` to `symbolic_convert.py:InstructionTranslatorBase.__init__`. In this PR, we use a global.~

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @ipiszy

Differential Revision: [D49257108](https://our.internmc.facebook.com/intern/diff/D49257108)

### CODE REVIEW

**Comment by ezyang** [comment]
Yeah I don't buy that this makes sense at all. The problem is that there isn't any reason that a meta that is valid on one node is valid for eg a decomposition. Take meta['val'] as a simple example: it says what the output size of a node is, which has no correspondence at all to the internal computation of the op!

**Comment by ezyang** [comment]
I guess what I don't really understand is, if you have a GraphModule, why do we have to go through all the rigamarole of Dynamo bytecode interpretation? Seems like you just want to ingest the GraphModule directly, which should basically be OK to go straight to aot export.

File: torch/_export/passes/add_runtime_assertions_for_constraints_pass.py [hunk 1]
```diff
@@ -141,7 +141,7 @@ def call(self, graph_module):
 
         # Populate the stack trace with dummy vals to respect IR
         for node in val.graph_module.graph.nodes:
-            if not hasattr(node.meta, "stack_trace"):
+            if False and not hasattr(node.meta, "stack_trace"):
```
Comment (ezyang) [inline torch/_export/passes/add_runtime_assertions_for_constraints_pass.py:144]: TODO I suppose

File: torch/_dynamo/eval_frame.py [hunk 2]
```diff
@@ -269,12 +269,16 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
-    def __call__(self, fn):
+    def __call__(self, fn, fx_node_meta_context=None):
         # public api for compiler config/options
         def get_compiler_config():
             return self.compiler_config
 
         fn = innermost_fn(fn)
+
+        if isinstance(fn, torch.fx.GraphModule):
+            fn.forward.__func__._torchdynamo_context = {"orig_graphmodule": fn}
```
Comment (ezyang) [inline torch/_dynamo/eval_frame.py:280]: If we inline into a callable corresponding to a GraphModule we should probably still be able to reestablish this context. This makes the tracking robust to changes on the enclosing code

File: torch/fx/graph.py [hunk 3]
```diff
@@ -588,12 +597,27 @@ def emit_node(node : Node):
 
         code = ''.join(body).lstrip('\n')
         code = '\n'.join('    ' + line for line in code.split('\n'))
+
+        # remove counter and generate lineno to node index mapping
```
Comment (ezyang) [inline torch/fx/graph.py:601]: It's not clear to me why you need to remove the counter lol

**Comment by ezyang** [comment]
This seems like it basically would work, but I haven't done a close mechanical review yet. There are some policy questions I have though. One big question I have is whether or not we should unconditionally trust the GraphModule metadata, or if we should also keep the "ground truth" actual metadata in some situations, e.g., for debugging purposes.

File: torch/fx/graph.py (hunk 3 above)
Comment (williamwen42) [reply torch/fx/graph.py:601]: To keep the GraphModule code clean

**Review by jansel — CHANGES_REQUESTED** [review]
Fix test failures

File: torch/_dynamo/code_context.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    def get_context(self, code: types.CodeType):
+        ctx = self.code_context.get(code)
+        if ctx is None:
+            ctx = {}
+            self.code_context[code] = ctx
+        return ctx
+
+    def pop_context(self, code: types.CodeType):
+        ctx = self.get_context(code)
+        self._remove_id(id(code))
```
Comment (jansel) [inline torch/_dynamo/code_context.py:22]: Where is `self._remove_id` defined?

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/code_context.py (hunk 4 above)
Comment (williamwen42) [reply torch/_dynamo/code_context.py:22]: typo

File: torch/_dynamo/eval_frame.py (hunk 2 above)
Comment (williamwen42) [reply torch/_dynamo/eval_frame.py]: done

**Comment by DanilBaibak** [comment]
Hi @williamwen42! Sorry, I have to revert your PR. It leads to internal failures. Here you can find more details - [D49190919](https://www.internalfb.com/diff/D49190919)

**Comment by williamwen42** [comment]
@williamwen42 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49257108).

**Comment by williamwen42** [comment]
Retrying - status of internal tests: D49257108

**Comment by williamwen42** [comment]
@williamwen42 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49257108).

**Comment by williamwen42** [comment]
@williamwen42 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D49257108).

**Comment by hyeokjunejeon** [comment]
I have been longing for this feature from the perspective of creating a Model Profile. Prior to this PR, it was impossible to determine which Python source code corresponded to the complex aten graph.

Thanks @williamwen42 for making this possible!

**Pull request closed without merging**